File: Data Structures & Algorithms/island-perimeter/submission-0.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    islandPerimeter(grid) {
        let perimeter = 0;
        for (let height = 0; height < grid.length; height++) {
            const current = grid[height];
            for (let width = 0; width < current.length; width++) {
                perimeter += this.getPerimeter(grid, width, height);
            }
        }
        return perimeter;
    }

    getPerimeter(grid, width, height) {
        // use row-major indexing: grid[row][col] -> grid[height][width]
        if (grid[height][width] === 0) {
            return 0;
        }

        let openSides = 0;
        // left
        if (width - 1 < 0 || grid[height][width - 1] === 0) openSides++;
        // up
        if (height - 1 < 0 || grid[height - 1][width] === 0) openSides++;
        // right
        if (width + 1 >= grid[0].length || grid[height][width + 1] === 0) openSides++;
        // down
        if (height + 1 >= grid.length || grid[height + 1][width] === 0) openSides++;

        return openSides;
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-0.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        def dfs(row: int, col: int):
            if col < 0 or col >= len(grid[0]):
                return 0
            
            if row < 0 or row >= len(grid):
                return 0

            if grid[row][col] == 0:
                return 0

            grid[row][col] = 0
            top =  dfs(row - 1, col)
            bottom = dfs(row + 1, col)
            left = dfs(row, col - 1)
            right = dfs(row, col + 1)

            return top + bottom + left + right + 1

        maxarea = 0

        for rowidx in range(len(grid)):
            row = grid[rowidx]
            for colidx in range(len(row)):
                col = row[colidx]

                if col == 0:
                    continue

                maxarea = max(maxarea, dfs(rowidx, colidx))

        return maxarea

        
        